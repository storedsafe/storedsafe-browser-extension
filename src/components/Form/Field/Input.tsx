import * as React from 'react';
import { FieldProps, FieldAttributes } from './Field';
import './Input.scss';

type InputProps = Pick<FieldProps, 'name' | 'type' | 'value' | 'onChange' | 'onBlur' | 'label'>;
export const Input: React.SFC<InputProps | FieldAttributes> = ({
  name,
  type,
  value,
  label,
  onChange,
  onBlur,
  ...attributes
}: InputProps) => (
  <div className={`field field-input field-input-${type}`}>
    <label htmlFor={name}>
      <span className="field-label">{label}</span>
      <div className="field-container">
        <input
          className="field-control"
          id={name}
          name={name}
          type={type}
          value={['number', 'range'].includes(type) && isNaN(value as number) ? '' : value as string | number}
          onChange={({ target }: React.ChangeEvent<HTMLInputElement>): void => {
            if (['number', 'range'].includes(type)) {
              onChange && onChange(target.valueAsNumber, name);
            } else {
              onChange && onChange(target.value, name);
            }
          }}
          onBlur={({ target }: React.FocusEvent<HTMLInputElement>): void => {
            if (['number', 'range'].includes(type)) {
              onBlur && onBlur(target.valueAsNumber, name);
            } else {
              onBlur && onBlur(target.value, name);
            }
          }}
          {...attributes}
        />
      </div>
    </label>
  </div>
);

Input.defaultProps = {
  type: 'text',
  value: '',
}
