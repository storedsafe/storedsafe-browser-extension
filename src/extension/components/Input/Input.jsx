/* eslint react/jsx-props-no-spreading: 0 */
import React from 'react';
import PropTypes from 'prop-types';

function Input({
  title,
  type,
  name,
  children,
  ...attrs
}) {
  if (type === 'checkbox') {
    return (
      <label htmlFor={name} className={type}>
        <span>{title}</span>
        <input
          type={type}
          name={name}
          id={name}
          {...attrs}
        />
        <span className="checkmark" />
      </label>
    );
  }

  if (type === 'select') {
    return (
      <label htmlFor={name} className={type}>
        <span>{title}</span>
        <div>
          <select
            name={name}
            id={name}
            {...attrs}
          >
            {children}
          </select>
        </div>
      </label>
    );
  }

  if (type === 'button' || type === 'submit') {
    return (
      <input
        className={type}
        type={type}
        name={name}
        id={name}
        value={title}
        {...attrs}
      />
    );
  }

  if (type === 'textarea') {
    return (
      <label htmlFor={name} className={type}>
        <span>{title}</span>
        <textarea
          name={name}
          id={name}
          {...attrs}
        />
      </label>
    );
  }

  return (
    <label htmlFor={name} className={type}>
      <span>{title}</span>
      <input
        type={type}
        name={name}
        id={name}
        {...attrs}
      />
    </label>
  );
}

Input.defaultProps = {
  children: null,
};

Input.propTypes = {
  title: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  children: PropTypes.node,
};

export default Input;
