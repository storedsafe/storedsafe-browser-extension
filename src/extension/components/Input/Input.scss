@use '../../base.scss';

/* Flow Styling */
label,
input[type='submit'],
input[type='button'] {
  margin-bottom: 1em;
}

label {
  display: flex;
  flex-direction: column;
  width: 100%;

  span {
    font-family: 'Montserrat', Helvetica, sans-serif;
    text-transform: uppercase;
  }
}

/* Component Styling */
input,
button,
select {
  border: 0;
  box-sizing: border-box;
  outline: 0;
  padding: 1em;
}

input:not([type=submit]):not([type=button]):not([type=checkbox]):not([type=radio]),
select,
option {
  background-color: base.$bg-light;
  border-bottom: 1px solid base.$border;
  color: #000;
  font-family: sans-serif;
  font-size: 1em;
  transition: base.$animation-duration ease;

  &:hover,
  &:focus {
    border-bottom-color: base.$accent;
  }

  &:focus {
    background-color: base.$bg-medium;
  }
}

option {
  // Chrome option inherits from select, firefox doesn't
  background-color: base.$bg-light;
}

button,
input[type=button],
input[type=submit] {
  background: base.$primary;
  border: 1px solid base.$bg-light;
  border-radius: base.$border-radius;
  color: base.$fg-light;
  font-weight: bold;
  text-transform: uppercase;
  transition: base.$animation-duration ease;
  width: 100%;

  &:hover,
  &:focus,
  &:active {
    background-color: base.$bg-light;
    border: 1px solid base.$accent;
    color: base.$accent;
  }

  &:active {
    background-color: base.$bg-medium;
  }
}

/* Disable and replace default style */
.checkbox {
  align-items: center;
  cursor: pointer;
  flex-direction: row-reverse;
  justify-content: flex-end;

  // Disable default checkbox
  input {
    height: 0;
    opacity: 0;
    position: absolute;
    width: 0;
  }

  // Create custom checkbox
  .checkmark {
    border: 1px solid base.$border;
    border-radius: base.$border-radius;
    height: 1.5em;
    margin-right: 0.5em;
    transition: base.$animation-duration ease;
    width: 1.5em;
  }

  // Create checkmark
  .checkmark::after {
    border: solid base.$fg-light;
    border-width: 0 0.2em 0.2em 0;
    box-sizing: border-box;
    // Ensure consistent size across browsers
    // Firefox acts weirdly if using em units
    font-size: 1rem;
    height: 60%;
    // Top margin less to prefer being pushed up rather than
    // down with pixel errors
    margin: 19% 35%;
    transform: rotate(45deg);
    width: 30%;
  }

  // Hide checkmark if not checked
  .checkmark::after {
    content: '';
    display: none;
  }

  // Show checkmark if checked
  input:checked ~ .checkmark::after {
    display: block;
  }

  // General appearance based on state
  &:hover input ~ .checkmark {
    border: 1px solid base.$accent;
  }

  input:focus ~ .checkmark {
    background-color: base.$bg-medium;
    border-color: base.$accent;
  }

  &:active input ~ .checkmark {
    background-color: base.$bg-medium;
  }

  input:checked ~ .checkmark {
    background-color: base.$accent;
  }
}

select {
  -moz-appearance: none; // stylelint-disable-line
  -webkit-appearance: none; // stylelint-disable-line
  appearance: none;
  padding-right: 2.25em;
  width: 100%;
}

.select div {
  position: relative;

  &::after {
    border: solid transparent;
    border-radius: base.$border-radius;
    border-top-color: base.$bg-dark;
    border-width: 0.7em 0.5em;
    box-sizing: border-box;
    content: '';
    height: 1em;
    margin-top: -0.35em;
    position: absolute;
    right: 0.75em;
    top: 50%;
    transition: base.$animation-duration;
    width: 1em;
  }
}
