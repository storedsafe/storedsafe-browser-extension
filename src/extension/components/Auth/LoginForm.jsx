import React, { useState } from 'react';
import PropTypes from 'prop-types';
import Input, { useForm } from '../Input';

/**
 * Enum containing the available login types.
 * */
const LoginType = Object.freeze({
  YUBIKEY: 'yubikey',
  TOTP: 'totp',
});

const LoginTypeData = Object.freeze({
  [LoginType.YUBIKEY]: {
    title: 'YubiKey',
    value: 'yubikey',
    fields: (inputs, onChange) => [
      <Input
        key="keys"
        type="password"
        name="keys"
        title="Passphrase + YubiKey"
        value={inputs.keys || ''}
        onChange={onChange}
        pattern=".+.{44}"
        required
      />,
    ],
    onSubmit: ({ keys }) => ({
      passphrase: keys.slice(0, -44),
      otp: keys.slice(-44),
    }),
  },
  [LoginType.TOTP]: {
    title: 'TOTP',
    value: 'totp',
    fields: (inputs, onChange) => [
      <Input
        key="passphrase"
        type="password"
        name="passphrase"
        title="Passphrase"
        value={inputs.passphrase || ''}
        onChange={onChange}
        required
      />,
      <Input
        key="otp"
        type="password"
        name="otp"
        title="TOTP"
        value={inputs.otp || ''}
        onChange={onChange}
        required
      />,
    ],
    onSubmit: ({ passphrase, otp }) => ({
      passphrase,
      otp,
    }),
  },
});

function LoginForm({
  defaultLoginType,
  onLogin,
}) {
  const [loginType, setLoginType] = useState(defaultLoginType);
  const { inputs, onChange } = useForm();

  // Change password fields when login type changes
  const onLoginTypeChange = (event) => {
    setLoginType(event.target.value);
  };

  // Create login type options from enum
  const loginOptions = Object.values(LoginType).map((key) => (
    <option key={key} value={LoginTypeData[key].value}>
      {LoginTypeData[key].title}
    </option>
  ));

  const onSubmit = (event) => {
    event.preventDefault();
    const { remember, username } = inputs;
    onLogin(loginType, remember, {
      username,
      ...LoginTypeData[loginType].onSubmit(inputs),
    });
  };

  return (
    <form className="login" onSubmit={onSubmit}>
      <Input
        type="select"
        name="logintype"
        title="Login Type"
        onChange={onLoginTypeChange}
      >
        {loginOptions}
      </Input>
      <Input
        type="text"
        name="username"
        title="Username or e-mail"
        value={inputs.username || ''}
        onChange={onChange}
        required
      />
      {/* Password fields depending on login type */}
      {LoginTypeData[loginType].fields(inputs, onChange)}
      <Input
        type="checkbox"
        name="remember"
        title="Remember Username"
        checked={inputs.remember || false}
        onChange={onChange}
      />
      <Input
        type="submit"
        name="submit"
        title="Login to StoredSafe"
      />
    </form>
  );
}

LoginForm.defaultProps = {
  defaultLoginType: LoginType.YUBIKEY,
};

LoginForm.propTypes = {
  defaultLoginType: PropTypes.string,
  onLogin: PropTypes.func.isRequired,
};

export { LoginType };
export default LoginForm;
