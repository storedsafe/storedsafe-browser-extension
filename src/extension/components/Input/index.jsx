import Input from './Input';
import useForm from './InputHooks';
import './Input.scss';

export { useForm };
export default Input;
