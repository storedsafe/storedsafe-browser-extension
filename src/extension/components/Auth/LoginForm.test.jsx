import React from 'react';
import { shallow, mount } from 'enzyme';
import LoginForm, { LoginType } from './LoginForm';
import Input from '../Input';

const onLogin = jest.fn();

describe('rendering', () => {
  test('<LoginForm />', () => {
    const wrapper = shallow(<LoginForm onLogin={onLogin} />);
    expect(wrapper.find('Input[name="logintype"]').length).toBe(1);
    expect(wrapper.find('Input[name="username"]').length).toBe(1);
    expect(wrapper.find('Input[name="keys"]').length).toBe(1);
    expect(wrapper.find('Input[name="remember"]').length).toBe(1);
    expect(wrapper.find('Input[type="submit"]').length).toBe(1);
    expect(wrapper.find(Input).length).toBe(5);
    expect(wrapper).toMatchSnapshot();
  });

  test('<LoginForm defaultLoginType="yubikey" />', () => {
    const wrapper = shallow((
      <LoginForm defaultLoginType={LoginType.YUBIKEY} onLogin={onLogin} />
    ));
    expect(wrapper.find('Input[name="logintype"]').length).toBe(1);
    expect(wrapper.find('Input[name="username"]').length).toBe(1);
    expect(wrapper.find('Input[name="keys"]').length).toBe(1);
    expect(wrapper.find('Input[name="remember"]').length).toBe(1);
    expect(wrapper.find('Input[type="submit"]').length).toBe(1);
    expect(wrapper.find(Input).length).toBe(5);
    expect(wrapper).toMatchSnapshot();
  });

  test('<LoginForm defaultLoginType="totp" />', () => {
    const wrapper = shallow((
      <LoginForm defaultLoginType={LoginType.TOTP} onLogin={onLogin} />
    ));
    expect(wrapper.find('Input[name="logintype"]').length).toBe(1);
    expect(wrapper.find('Input[name="username"]').length).toBe(1);
    expect(wrapper.find('Input[name="passphrase"]').length).toBe(1);
    expect(wrapper.find('Input[name="otp"]').length).toBe(1);
    expect(wrapper.find('Input[name="remember"]').length).toBe(1);
    expect(wrapper.find('Input[type="submit"]').length).toBe(1);
    expect(wrapper.find(Input).length).toBe(6);
    expect(wrapper).toMatchSnapshot();
  });
});

describe('interaction', () => {
  test('<LoginForm />, changes fields on change loginType', () => {
    const wrapper = mount((
      <LoginForm defaultLoginType={LoginType.YUBIKEY} onLogin={onLogin} />
    ));
    const select = wrapper.find('select');
    expect(wrapper.find('input[name="keys"]').length).toBe(1);
    expect(wrapper.find('input[name="passphrase"]').length).toBe(0);
    expect(wrapper.find('input[name="otp"]').length).toBe(0);
    select.simulate('change', { target: { value: LoginType.TOTP } });
    expect(wrapper.find('input[name="keys"]').length).toBe(0);
    expect(wrapper.find('input[name="passphrase"]').length).toBe(1);
    expect(wrapper.find('input[name="otp"]').length).toBe(1);
  });

  test('<LoginForm defaultLoginType="yubikey" />, submit', () => {
    const username = 'fredrik';
    const passphrase = 's3cr3t';
    const otp = 'cccccccccccccccccccccccccccccccccccccccccccc';
    const wrapper = mount((
      <LoginForm defaultLoginType={LoginType.YUBIKEY} onLogin={onLogin} />
    ));
    wrapper.find('input[name="username"]')
      .simulate('change', {
        target: { name: 'username', value: username },
      });
    wrapper.find('input[name="keys"]')
      .simulate('change', {
        target: { name: 'keys', value: passphrase + otp },
      });
    wrapper.find('input[name="remember"]')
      .simulate('change', {
        target: { type: 'checkbox', name: 'remember', checked: true },
      });
    wrapper.simulate('submit');
    expect(onLogin).toHaveBeenLastCalledWith(
      LoginType.YUBIKEY,
      true,
      { username, passphrase, otp },
    );
  });

  test('<LoginForm defaultLoginType="totp" />, submit', () => {
    const username = 'peter';
    const passphrase = 'p4ssw0rd';
    const otp = '123987';
    const wrapper = mount((
      <LoginForm defaultLoginType={LoginType.TOTP} onLogin={onLogin} />
    ));
    wrapper.find('input[name="username"]')
      .simulate('change', {
        target: { name: 'username', value: username },
      });
    wrapper.find('input[name="passphrase"]')
      .simulate('change', {
        target: { name: 'passphrase', value: passphrase },
      });
    wrapper.find('input[name="otp"]')
      .simulate('change', {
        target: { name: 'otp', value: otp },
      });
    wrapper.find('input[name="remember"]')
      .simulate('change', {
        target: { type: 'checkbox', name: 'remember', checked: false },
      });
    wrapper.simulate('submit');
    expect(onLogin).toHaveBeenLastCalledWith(
      LoginType.TOTP,
      false,
      { username, passphrase, otp },
    );
  });
});
