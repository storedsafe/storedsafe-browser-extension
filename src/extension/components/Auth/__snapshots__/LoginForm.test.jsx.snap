// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering <LoginForm /> 1`] = `ShallowWrapper {}`;

exports[`rendering <LoginForm defaultLoginType="totp" /> 1`] = `ShallowWrapper {}`;

exports[`rendering <LoginForm defaultLoginType="yubikey" /> 1`] = `ShallowWrapper {}`;
